Crash in sandvoxed google play

```
type: crash
osVersion: google/oriole/oriole:13/TP1A.221005.002/2022100300:user/release-keys
package: com.google.android.gms:222615044
process: com.google.android.gms.persistent

java.lang.SecurityException: query intent receivers: UID 1010135 requires android.permission.INTERACT_ACROSS_USERS_FULL or android.permission.INTERACT_ACROSS_USERS to access user 0.
	at android.os.Parcel.createExceptionOrNull(Parcel.java:3021)
	at android.os.Parcel.createException(Parcel.java:3005)
	at android.os.Parcel.readException(Parcel.java:2981)
	at android.os.Parcel.readException(Parcel.java:2923)
	at android.bluetooth.IBluetoothManager$Stub$Proxy.bindBluetoothProfileService(IBluetoothManager.java:500)
	at android.bluetooth.BluetoothHeadset.doBind(BluetoothHeadset.java:407)
	at android.bluetooth.BluetoothHeadset.<init>(BluetoothHeadset.java:398)
	at android.bluetooth.BluetoothAdapter.getProfileProxy(BluetoothAdapter.java:3628)
	at crpw.<init>(:com.google.android.gms@222615044@22.26.15 (190400-461192076):7)
	at com.google.android.gms.nearby.fastpair.service.WearableDataListenerChimeraService.m(:com.google.android.gms@222615044@22.26.15 (190400-461192076):2)
	at com.google.android.gms.nearby.fastpair.service.WearableDataListenerChimeraService.r(:com.google.android.gms@222615044@22.26.15 (190400-461192076):1)
	at auxk.run(:com.google.android.gms@222615044@22.26.15 (190400-461192076):0)
	at xot.c(:com.google.android.gms@222615044@22.26.15 (190400-461192076):6)
	at xot.run(:com.google.android.gms@222615044@22.26.15 (190400-461192076):7)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
	at xty.run(:com.google.android.gms@222615044@22.26.15 (190400-461192076):0)
	at java.lang.Thread.run(Thread.java:1012)
	Suppressed: cfvr: 
		at tk_trace.nearby-WearableDataListenerChimeraService_onCreate(Unknown Source:0)
		at tk_trace.AccountManager.addOnAccountsUpdatedListener(Unknown Source:0)
		at tk_trace.cast-CastChimeraService_onCreate(Unknown Source:0)
		at tk_trace.container-CastPersistentBoundBrokerChimeraService_onBind(Unknown Source:0)
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.pm.ComputerEngine.enforceCrossUserOrProfilePermission(ComputerEngine.java:2960)
	at com.android.server.pm.ComputerEngine.queryIntentServicesInternal(ComputerEngine.java:613)
	at com.android.server.pm.IPackageManagerBase.queryIntentServices(IPackageManagerBase.java:1061)
	at android.app.ApplicationPackageManager.queryIntentServicesAsUser(ApplicationPackageManager.java:1599)
	at android.app.ApplicationPackageManager.queryIntentServices(ApplicationPackageManager.java:1620)


```

--- a/gmscompat_config
+++ b/gmscompat_config
@@ -16,6 +16,10 @@ PromoFeature__enabled perms-none-of android.permission.ACCESS_COARSE_LOCATION fa
 # controls access to privileged StatsManager#getRegisteredExperimentIds()
 add_external_experiment_ids false
 
+[com.google.android.gms.nearby]
+# disable WearableDataListenerService, it needs privileged permissions
+FastPairFeature__enable_wearable_service false
+
 [com.google.android.gms.cast]
 # Prevent crashes in GmsCore caused by calling ConnectivityManager#registerNetworkCallback too many
 # times when it has no INTERNET permission (there's a OS-enforced limit on the number of these callbacks).
@@ -73,6 +77,7 @@ getName nullString
 getAddress String 02:00:00:00:00:00
 listenUsingInsecureL2capChannel throw java.io.IOException
 disable false
+getProfileProxy false
 
 [android.bluetooth.BluetoothDevice]
 getMetadata default

--- a/src/app/grapheneos/gmscompat/location/Client.kt
+++ b/src/app/grapheneos/gmscompat/location/Client.kt
@@ -2,11 +2,8 @@ package app.grapheneos.gmscompat.location
 
 import android.Manifest
 import android.app.AppOpsManager
-import android.content.AttributionSource
 import android.content.Context
-import android.content.ContextParams
 import android.content.pm.PackageManager.PERMISSION_GRANTED
-import android.location.Criteria
 import android.location.LocationManager
 import android.os.Binder
 import androidx.annotation.CheckResult
@@ -59,37 +56,6 @@ class Client(val gls: GLocationService, unverifiedPackageName: String? = null, v
     }
 
     val locationManager = ctx.getSystemService(LocationManager::class.java)!!
-
-    fun getProvider(req: LocationRequest): String {
-        val priority = req.priority
-        if (priority == LocationRequest.PRIORITY_NO_POWER) {
-            return LocationManager.PASSIVE_PROVIDER
-        } else {
-            return LocationManager.GPS_PROVIDER
-        }
-        // getBestProvider() returns FUSED_PROVIDER sometimes, which doesn't do anything useful when
-        // there's no NETWORK_PROVIDER, but exposes more potential framework bugs
-        /*
-        val criteria = Criteria()
-        when (priority) {
-            LocationRequest.PRIORITY_HIGH_ACCURACY ->
-                criteria.accuracy = Criteria.ACCURACY_FINE
-            LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY ->
-                criteria.accuracy = Criteria.ACCURACY_FINE
-            LocationRequest.PRIORITY_LOW_POWER ->
-                criteria.powerRequirement = Criteria.POWER_LOW
-            else ->
-                throw IllegalArgumentException(priority.toString())
-        }
-        return locationManager.getBestProvider(criteria, true)!!
-         */
-    }
-
-    fun enforceFinePermission() {
-        if (permission != Permission.FINE) {
-            throw SecurityException("missing " + Permission.FINE.string())
-        }
-    }
 
     @CheckResult
     fun noteProxyAppOp(): Int {

--- a/src/app/grapheneos/gmscompat/location/GLocationService.kt
+++ b/src/app/grapheneos/gmscompat/location/GLocationService.kt
@@ -2,7 +2,6 @@ package app.grapheneos.gmscompat.location
 
 import android.app.AppOpsManager
 import android.app.AppOpsManager.MODE_ALLOWED
-import android.location.Criteria
 import android.location.GnssAntennaInfo
 import android.location.GnssMeasurementsEvent
 import android.location.GnssNavigationMessage
@@ -10,10 +9,12 @@ import android.location.GnssStatus
 import android.location.Location
 import android.location.LocationManager
 import android.location.OnNmeaMessageListener
-import android.location.provider.ProviderProperties
 import android.os.Binder
+import android.os.CancellationSignal
+import android.os.RemoteException
 import android.util.SparseArray
 import com.google.android.gms.common.api.Status
+import com.google.android.gms.common.internal.ICancelToken
 import com.google.android.gms.location.ILocationCallback
 import com.google.android.gms.location.ILocationListener
 import com.google.android.gms.location.LocationAvailability
@@ -24,7 +25,10 @@ import com.google.android.gms.location.LocationSettingsStates
 import com.google.android.gms.location.internal.FusedLocationProviderResult
 import com.google.android.gms.location.internal.IFusedLocationProviderCallback
 import com.google.android.gms.location.internal.IGoogleLocationManagerService
+import com.google.android.gms.location.internal.ILocationAvailabilityStatusCallback
+import com.google.android.gms.location.internal.ILocationStatusCallback
 import com.google.android.gms.location.internal.ISettingsCallbacks
+import com.google.android.gms.location.internal.IStatusCallback
 import com.google.android.gms.location.internal.LocationRequestUpdateData
 import app.grapheneos.gmscompat.App
 import app.grapheneos.gmscompat.Const
@@ -32,9 +36,14 @@ import app.grapheneos.gmscompat.logd
 import app.grapheneos.gmscompat.objectToString
 import app.grapheneos.gmscompat.opModeToString
 import com.google.android.gms.common.api.CommonStatusCodes
+import com.google.android.gms.location.CurrentLocationRequest
+import com.google.android.gms.location.LastLocationRequest
+import com.google.android.gms.location.LocationAvailabilityRequest
+import com.google.android.gms.location.internal.LocationReceiver
 import java.lang.ref.WeakReference
 import java.util.Arrays
 import java.util.concurrent.Executors
+import java.util.function.Consumer
 
 object GLocationService : IGoogleLocationManagerService.Stub() {
     val ctx = App.ctx()
@@ -47,6 +56,75 @@ object GLocationService : IGoogleLocationManagerService.Stub() {
     // gc would be hard to do correctly if listeners were strongly referenced
     val mapOfListeners = SparseArray<WeakReference<Listeners>>(50)
 
+    override fun registerLocationReceiver(receiver: LocationReceiver, request: LocationRequest, callback: IStatusCallback) {
+        val client = Client(this)
+
+//        logd{client.packageName}
+
+        val clientListeners = getOrCreateListeners(client)
+
+        val key: Any
+        val glf: GLocationForwarder
+        when (receiver.type) {
+            LocationReceiver.TYPE_ILocationListener -> {
+                key = receiver.binder
+                glf = GlfLocationListener(ILocationListener.Stub.asInterface(receiver.binder))
+            }
+            LocationReceiver.TYPE_ILocationCallback -> {
+                key = receiver.binder
+                glf = GlfLocationCallback(ILocationCallback.Stub.asInterface(receiver.binder))
+            }
+            LocationReceiver.TYPE_PendingIntent -> {
+                key = receiver.pendingIntent
+                glf = GlfPendingIntent(receiver.pendingIntent)
+            }
+            else -> {
+                throw IllegalArgumentException()
+            }
+        }
+        glf.listeners = clientListeners
+
+        val provider = OsLocationProvider.get(client, request)
+
+        val oll = OsLocationListener(client, provider, request.toOsLocationRequest(), glf)
+        clientListeners.update(client, key, oll)
+
+        callback.onCompletion(Status.SUCCESS)
+    }
+
+    override fun unregisterLocationReceiver(receiver: LocationReceiver, callback: IStatusCallback) {
+        val client = Client(this)
+
+        val clientListeners = getOrCreateListeners(client)
+
+        val key: Any = when (receiver.type) {
+            LocationReceiver.TYPE_ILocationListener,
+            LocationReceiver.TYPE_ILocationCallback ->
+                receiver.binder
+            LocationReceiver.TYPE_PendingIntent ->
+                receiver.pendingIntent
+            else -> throw IllegalArgumentException()
+        }
+
+//        logd{"${client.packageName} key ${key}"}
+
+        clientListeners.remove(client, key)
+        callback.onCompletion(Status.SUCCESS)
+    }
+
+    private fun getOrCreateListeners(client: Client): Listeners {
+        return synchronized(mapOfListeners) {
+            val cur = mapOfListeners.get(client.uid)?.get()
+            if (cur != null) {
+                cur
+            } else {
+                val l = Listeners(7)
+                mapOfListeners.put(client.uid, WeakReference(l))
+                l
+            }
+        }
+    }
+
     override fun updateLocationRequest(data: LocationRequestUpdateData) {
 //        logd{data}
         val pendingIntent = data.pendingIntent
@@ -58,17 +136,8 @@ object GLocationService : IGoogleLocationManagerService.Stub() {
             return
         }
 
-        val listeners: Listeners =
-        synchronized(mapOfListeners) {
-            val cur = mapOfListeners.get(client.uid)?.get()
-            if (cur != null) {
-                cur
-            } else {
-                val l = Listeners(7)
-                mapOfListeners.put(client.uid, WeakReference(l))
-                l
-            }
-        }
+        val clientListeners = getOrCreateListeners(client)
+
         if (data.opCode == LocationRequestUpdateData.OP_REQUEST_UPDATES) {
             val key: Any
             val glf: GLocationForwarder
@@ -89,11 +158,12 @@ object GLocationService : IGoogleLocationManagerService.Stub() {
                     glf = GlfLocationListener(llistener)
                 }
             }
-            glf.listeners = listeners
+            glf.listeners = clientListeners
 
             val req: LocationRequest = data.request.request
-            val oll = OsLocationListener(client, client.getProvider(req), req.toOsLocationRequest(), glf)
-            listeners.update(client, key, oll)
+            val oll = OsLocationListener(client, OsLocationProvider.get(client, req),
+                    req.toOsLocationRequest(), glf)
+            clientListeners.update(client, key, oll)
         } else {
             require(data.opCode == LocationRequestUpdateData.OP_REMOVE_UPDATES)
             val key: Any
@@ -108,7 +178,7 @@ object GLocationService : IGoogleLocationManagerService.Stub() {
                     key = llistener!!.asBinder()
                 }
             }
-            listeners.remove(client, key)
+            clientListeners.remove(client, key)
         }
         data.fusedLocationProviderCallback?.onFusedLocationProviderResult(FusedLocationProviderResult.SUCCESS)
     }
@@ -163,25 +233,43 @@ object GLocationService : IGoogleLocationManagerService.Stub() {
             }
             return null
         }
-        val criteria = Criteria()
-        criteria.accuracy =
-        if (client.permission == Permission.FINE) {
-            Criteria.ACCURACY_FINE
-        } else {
-            Criteria.ACCURACY_COARSE
+
+        val provider = OsLocationProvider.get(client, LocationRequest.GRANULARITY_PERMISSION_LEVEL)
+
+        return client.locationManager.getLastKnownLocation(provider.name)?.let {
+            provider.maybeFudge(it)
         }
-        val lm = client.locationManager
-        val provider = lm.getBestProvider(criteria, true)!!
-        val loc = lm.getLastKnownLocation(provider)
-        if (loc != null) {
-            if (client.permission == Permission.COARSE) {
-                val pp = lm.getProviderProperties(provider)
-                if (pp!!.accuracy == ProviderProperties.ACCURACY_FINE) {
-                    return LocationFudger().createCoarse(loc)
-                }
+    }
+
+    override fun getLastLocation4(request: LastLocationRequest, callback: ILocationStatusCallback) {
+        val client = Client(this)
+        logd{"client ${client.packageName} maxAge ${request.maxAge}"}
+
+        val provider = OsLocationProvider.get(client, request.granularity)
+
+        var location = client.locationManager.getLastKnownLocation(provider.name)?.let {
+            if (it.elapsedRealtimeAgeMillis <= request.maxAge) {
+                provider.maybeFudge(it)
+            } else {
+                null
             }
         }
-        return loc
+
+        if (location != null) {
+            val opMode = client.noteProxyAppOp()
+            if (opMode != MODE_ALLOWED) {
+                logd{"opMode ${opModeToString(opMode)}"}
+                location = null
+                // GmsCore returns Status.SUCCESS even in this case
+            }
+        }
+
+        callback.onResult(Status.SUCCESS, location)
+    }
+
+    override fun getLastLocation5(request: LastLocationRequest, receiver: LocationReceiver) {
+        require(receiver.type == LocationReceiver.TYPE_ILocationStatusCallback)
+        getLastLocation4(request, ILocationStatusCallback.Stub.asInterface(receiver.binder))
     }
 
     override fun getLastLocation(): Location? {
@@ -195,6 +283,64 @@ object GLocationService : IGoogleLocationManagerService.Stub() {
         return getLastLocation3(null)
     }
 
+    override fun getCurrentLocation(request: CurrentLocationRequest, callback: ILocationStatusCallback): ICancelToken {
+        val client = Client(this)
+
+        logd{client.packageName}
+
+        val osRequest = android.location.LocationRequest.Builder(0L).apply {
+            setDurationMillis(request.durationMillis)
+            setQuality(gmsPriorityToOsQuality(request.priority))
+        }.build()
+
+        val provider = OsLocationProvider.get(client, request.priority, request.granularity)
+
+        val consumer = Consumer<Location> { origLocation ->
+            // location.elapsedRealtimeAgeMillis <= request.maxUpdateAgeMillis
+            // check is not needed, maxUpdateAgeMillis applies only to historical locations
+            // which are never returned by OsLocationProvider
+            if (origLocation != null) {
+                var location: Location? = provider.maybeFudge(origLocation)
+
+                val opMode = client.noteProxyAppOp()
+                if (opMode != MODE_ALLOWED) {
+                    logd{"opMode ${opModeToString(opMode)}"}
+                    // GmsCore returns Status.SUCCESS even in this case
+                    location = null
+                }
+
+                try {
+                    callback.onResult(Status.SUCCESS, location)
+                } catch (e: RemoteException) {
+                    logd{e}
+                }
+            } else {
+                try {
+                    callback.onResult(Status(CommonStatusCodes.TIMEOUT), null)
+                } catch (e: RemoteException) {
+                    logd{e}
+                }
+            }
+        }
+
+        val cancellationSignal = CancellationSignal()
+
+        client.locationManager.getCurrentLocation(provider.name, osRequest, cancellationSignal,
+            listenerCallbacksExecutor, consumer)
+
+        return object : ICancelToken.Stub() {
+            override fun cancel() {
+                logd{"cancel signal from ${client.packageName}"}
+                cancellationSignal.cancel()
+            }
+        }
+    }
+
+    override fun getCurrentLocation2(request: CurrentLocationRequest, receiver: LocationReceiver): ICancelToken {
+        require(receiver.type == LocationReceiver.TYPE_ILocationStatusCallback)
+        return getCurrentLocation(request, ILocationStatusCallback.Stub.asInterface(receiver.binder))
+    }
+
     // https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient#getLocationAvailability()
     override fun getLocationAvailability(packageName: String?): LocationAvailability {
         val client = try {
@@ -204,6 +350,14 @@ object GLocationService : IGoogleLocationManagerService.Stub() {
         }
         logd{"client ${client.packageName}"}
         return LocationAvailability.get(client.locationManager.isLocationEnabled())
+    }
+
+    override fun getLocationAvailability2(request: LocationAvailabilityRequest, receiver: LocationReceiver) {
+        val res = getLocationAvailability(null)
+
+        require(receiver.type == LocationReceiver.TYPE_ILocationAvailabilityStatusCallback)
+        val callback = ILocationAvailabilityStatusCallback.Stub.asInterface(receiver.binder)
+        callback.onResult(Status.SUCCESS, res)
     }
 
     // https://developers.google.com/android/reference/com/google/android/gms/location/SettingsClient#checkLocationSettings(com.google.android.gms.location.LocationSettingsRequest)
@@ -226,13 +380,23 @@ object GLocationService : IGoogleLocationManagerService.Stub() {
 
     @JvmField
     val CODES = intArrayOf(
-        FIRST_CALL_TRANSACTION + 58, // updateLocationRequest
-        FIRST_CALL_TRANSACTION + 66, // flushLocations
-        FIRST_CALL_TRANSACTION + 6,  // getLastLocation
-        FIRST_CALL_TRANSACTION + 20, // getLastLocation2
-        FIRST_CALL_TRANSACTION + 79, // getLastLocation3
-        FIRST_CALL_TRANSACTION + 33, // getLocationAvailability
-        FIRST_CALL_TRANSACTION + 62, // requestLocationSettingsDialog
+        // to generate, copy TRANSACTION_* definitions from IGoogleLocationManagerService.Stub and run:
+        // xsel --clipboard | sed -e 's/static final int TRANSACTION_/\/* /' -e 's/=/*\//' -e 's/;/,/' | xsel --clipboard --input
+
+        /* getLastLocation */ 7,
+        /* getLastLocation2 */ 21,
+        /* getLastLocation3 */ 80,
+        /* getLastLocation4 */ 82,
+        /* getLastLocation5 */ 90,
+        /* getLocationAvailability */ 34,
+        /* getLocationAvailability2 */ 91,
+        /* requestLocationSettingsDialog */ 63,
+        /* updateLocationRequest */ 59,
+        /* flushLocations */ 67,
+        /* registerLocationReceiver */ 88,
+        /* unregisterLocationReceiver */ 89,
+        /* getCurrentLocation */ 87,
+        /* getCurrentLocation2 */ 92,
     )
     init {
         Arrays.sort(CODES)

--- a/src/app/grapheneos/gmscompat/location/Listeners.kt
+++ b/src/app/grapheneos/gmscompat/location/Listeners.kt
@@ -40,7 +40,7 @@ fun Listeners.update(client: Client, key: Any, listener: OsLocationListener) {
             }
 
         }
-        client.locationManager.requestLocationUpdates(listener.provider, listener.request,
+        client.locationManager.requestLocationUpdates(listener.provider.name, listener.request,
             GLocationService.listenerCallbacksExecutor, listener)
 
         val curIdx = indexOfKey(key)

--- a/src/app/grapheneos/gmscompat/location/LocationFudger.java
+++ b/src/app/grapheneos/gmscompat/location/LocationFudger.java
@@ -31,7 +31,7 @@ import androidx.annotation.VisibleForTesting;
  * Contains the logic to obfuscate (fudge) locations for coarse applications. The goal is just to
  * prevent applications with only the coarse location permission from receiving a fine location.
  */
-class LocationFudger {
+public class LocationFudger {
 
     // minimum accuracy a coarsened location can have
     private static final float MIN_ACCURACY_M = 200.0f;

--- a/src/app/grapheneos/gmscompat/location/OsLocationListener.kt
+++ b/src/app/grapheneos/gmscompat/location/OsLocationListener.kt
@@ -10,9 +10,9 @@ import app.grapheneos.gmscompat.logd
 import app.grapheneos.gmscompat.opModeToString
 import com.google.android.gms.location.LocationAvailability
 import com.google.android.gms.location.LocationRequest
-import java.lang.IllegalStateException
 import java.util.Collections
 import java.util.concurrent.CountDownLatch
+import kotlin.IllegalStateException
 import kotlin.math.max
 import kotlin.math.min
 
@@ -53,20 +53,13 @@ fun gmsPriorityToOsQuality(priority: Int): Int =
             throw IllegalArgumentException()
     }
 
-class OsLocationListener(val client: Client, val provider: String,
+class OsLocationListener(val client: Client, val provider: OsLocationProvider,
                          val request: android.location.LocationRequest,
                          val forwarder: GLocationForwarder
 ) : LocationListener {
     init {
         forwarder.osLocationListener = this
     }
-
-    private val fudger =
-        if (client.permission == Permission.COARSE) {
-            LocationFudger()
-        } else {
-            null
-        }
 
     override fun onLocationChanged(location: Location) {
         onLocationChanged(Collections.singletonList(location))
@@ -84,14 +77,7 @@ class OsLocationListener(val client: Client, val provider: String,
             it.provider = LocationManager.FUSED_PROVIDER
         }
 
-        val locationsToForward =
-        if (fudger != null) {
-            locations.map {
-                fudger.createCoarse(it)
-            }
-        } else {
-            locations
-        }
+        val locationsToForward = locations.map { provider.maybeFudge(it) }
 
         if (Const.DEV
             && false
@@ -126,13 +112,13 @@ class OsLocationListener(val client: Client, val provider: String,
 
     override fun onProviderEnabled(provider: String) {
         logd{provider}
-        check(provider == this.provider)
+        check(provider == this.provider.name)
         onLocationAvailabilityChanged(true)
     }
 
     override fun onProviderDisabled(provider: String) {
         logd{provider}
-        check(provider == this.provider)
+        check(provider == this.provider.name)
         onLocationAvailabilityChanged(false)
     }
 
@@ -143,7 +129,7 @@ class OsLocationListener(val client: Client, val provider: String,
         synchronized(this) {
             flushLatch = l
             try {
-                client.locationManager.requestFlush(provider, this, 0)
+                client.locationManager.requestFlush(provider.name, this, 0)
             } catch (e: IllegalStateException) {
                 // may get thrown if other thread unregisters this listener
                 return

--- a/src/app/grapheneos/gmscompat/location/OsLocationProvider.kt
+++ b/src/app/grapheneos/gmscompat/location/OsLocationProvider.kt
@@ -1,0 +1,122 @@
+package app.grapheneos.gmscompat.location
+
+import android.location.Criteria
+import android.location.Location
+import android.location.LocationManager
+import android.location.provider.ProviderProperties
+import com.google.android.gms.location.LocationRequest
+
+class OsLocationProvider(val name: String, val properties: ProviderProperties?, val fudger: LocationFudger?) {
+
+    fun maybeFudge(location: Location): Location {
+        if (fudger == null) {
+            return location
+        }
+
+        return fudger.createCoarse(location)
+    }
+
+    companion object {
+        fun get(client: Client, req: LocationRequest) = get(client, req.priority, req.granularity)
+
+        fun get(client: Client, priority: Int, granularity: Int): OsLocationProvider {
+            val name = if (priority == LocationRequest.PRIORITY_NO_POWER) {
+                LocationManager.PASSIVE_PROVIDER
+            } else {
+                LocationManager.GPS_PROVIDER
+            }
+            // getBestProvider() returns FUSED_PROVIDER sometimes, which doesn't do anything useful when
+            // there's no NETWORK_PROVIDER, but exposes more potential framework bugs
+            /*
+            val criteria = Criteria()
+            when (priority) {
+                LocationRequest.PRIORITY_HIGH_ACCURACY ->
+                    criteria.accuracy = Criteria.ACCURACY_FINE
+                LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY ->
+                    criteria.accuracy = Criteria.ACCURACY_FINE
+                LocationRequest.PRIORITY_LOW_POWER ->
+                    criteria.powerRequirement = Criteria.POWER_LOW
+                else ->
+                    throw IllegalArgumentException(priority.toString())
+            }
+            return locationManager.getBestProvider(criteria, true)!!
+             */
+
+            return get(client, name, granularity)
+        }
+
+        fun get(client: Client, granularity: Int): OsLocationProvider {
+            // see the comment above
+            /*
+            val criteria = Criteria()
+            criteria.accuracy = when (client.permission) {
+                Permission.COARSE -> {
+                    when (granularity) {
+                        LocationRequest.GRANULARITY_PERMISSION_LEVEL,
+                        LocationRequest.GRANULARITY_COARSE ->
+                            Criteria.ACCURACY_COARSE
+                        LocationRequest.GRANULARITY_FINE -> throw SecurityException()
+                        else -> throw IllegalArgumentException()
+                    }
+                }
+                Permission.FINE -> {
+                    when (granularity) {
+                        LocationRequest.GRANULARITY_COARSE ->
+                            Criteria.ACCURACY_COARSE
+                        LocationRequest.GRANULARITY_PERMISSION_LEVEL,
+                        LocationRequest.GRANULARITY_FINE ->
+                            Criteria.ACCURACY_FINE
+                        else -> throw IllegalArgumentException()
+                    }
+                }
+            }
+
+            val name: String = client.locationManager.getBestProvider(criteria, true)!!
+             */
+            val name = LocationManager.GPS_PROVIDER
+
+            return get(client, name, granularity)
+        }
+
+        fun get(client: Client, name: String, granularity: Int): OsLocationProvider {
+            val properties: ProviderProperties? = client.locationManager.getProviderProperties(name)
+
+            val fudger: LocationFudger? = when (client.permission) {
+                Permission.COARSE -> {
+                    when (granularity) {
+                        LocationRequest.GRANULARITY_PERMISSION_LEVEL,
+                        LocationRequest.GRANULARITY_COARSE -> {
+                            if (properties == null) {
+                                LocationFudger()
+                            } else when (properties.accuracy) {
+                                ProviderProperties.ACCURACY_FINE -> LocationFudger()
+                                ProviderProperties.ACCURACY_COARSE -> null
+                                else -> throw IllegalStateException()
+                            }
+                        }
+                        LocationRequest.GRANULARITY_FINE -> throw SecurityException()
+                        else -> throw IllegalArgumentException()
+                    }
+                }
+                Permission.FINE -> {
+                    when (granularity) {
+                        LocationRequest.GRANULARITY_COARSE ->
+                            if (properties == null) {
+                                LocationFudger()
+                            } else when (properties.accuracy) {
+                                ProviderProperties.ACCURACY_FINE -> LocationFudger()
+                                ProviderProperties.ACCURACY_COARSE -> null
+                                else -> throw IllegalStateException()
+                            }
+                        LocationRequest.GRANULARITY_PERMISSION_LEVEL,
+                        LocationRequest.GRANULARITY_FINE ->
+                            null
+                        else -> throw IllegalArgumentException()
+                    }
+                }
+            }
+
+            return OsLocationProvider(name, properties, fudger)
+        }
+    }
+}

--- a/src/app/grapheneos/gmscompat/safeparcel/SafeParcel.java
+++ b/src/app/grapheneos/gmscompat/safeparcel/SafeParcel.java
@@ -722,7 +722,7 @@ public class SafeParcel {
         return res;
     }
 
-    public static <T extends Parcelable> void writeParcelableInternal(int pi, List<T> v, Parcel p, int flags) {
+    public static <T extends Parcelable> void writeParcelableList(int pi, List<T> v, Parcel p, int flags) {
         if (v != null) {
             int phe = beginPropHeader(pi, p);
             int size = v.size();

--- a/src/com/google/android/gms/common/internal/ICancelToken.aidl
+++ b/src/com/google/android/gms/common/internal/ICancelToken.aidl
@@ -1,0 +1,5 @@
+package com.google.android.gms.common.internal;
+
+interface ICancelToken {
+    void cancel() = 1;
+}

--- a/src/com/google/android/gms/location/CurrentLocationRequest.aidl
+++ b/src/com/google/android/gms/location/CurrentLocationRequest.aidl
@@ -1,0 +1,3 @@
+package com.google.android.gms.location;
+
+parcelable CurrentLocationRequest;

--- a/src/com/google/android/gms/location/CurrentLocationRequest.java
+++ b/src/com/google/android/gms/location/CurrentLocationRequest.java
@@ -1,0 +1,55 @@
+package com.google.android.gms.location;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import app.grapheneos.gmscompat.safeparcel.Property;
+import app.grapheneos.gmscompat.safeparcel.SafeParcel;
+import app.grapheneos.gmscompat.safeparcel.SpReadOnly;
+
+// https://developers.google.com/android/reference/com/google/android/gms/location/CurrentLocationRequest
+public class CurrentLocationRequest extends SpReadOnly {
+    @Property(1) public long maxUpdateAgeMillis = 60_000L;
+    @Property(2) public int granularity = LocationRequest.GRANULARITY_PERMISSION_LEVEL;
+    @Property(3) public int priority = LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY;
+    @Property(4) public long durationMillis = Long.MAX_VALUE;
+//    @Property(5) public boolean bypass;
+//    @Property(6) public WorkSource workSource;
+//    @Property(7) public int throttleBehavior;
+//    @Property(8) public String moduleId;
+//    @Property(9) public ClientIdentity impersonation;
+
+// SafeParcel code block generated with Spoon | START
+    public static final Parcelable.Creator<CurrentLocationRequest> CREATOR = new Parcelable.Creator<CurrentLocationRequest>() {
+        public CurrentLocationRequest createFromParcel(Parcel p) {
+            CurrentLocationRequest o = new CurrentLocationRequest();
+            final int objectEnd = SafeParcel.getObjectEnd(p);
+            while (p.dataPosition() < objectEnd) {
+                int ph = SafeParcel.propHeader(p);
+                switch (SafeParcel.propId(ph)) {
+                    case 1 :
+                        o.maxUpdateAgeMillis = SafeParcel.readLong(ph, p);
+                        continue;
+                    case 2 :
+                        o.granularity = SafeParcel.readInt(ph, p);
+                        continue;
+                    case 3 :
+                        o.priority = SafeParcel.readInt(ph, p);
+                        continue;
+                    case 4 :
+                        o.durationMillis = SafeParcel.readLong(ph, p);
+                        continue;
+                    default :
+                        SafeParcel.skipProp(ph, p);
+                }
+            }
+            SafeParcel.checkFullRead(objectEnd, p);
+            return o;
+        }
+
+        public CurrentLocationRequest[] newArray(int size) {
+            return new CurrentLocationRequest[size];
+        }
+    };
+// SafeParcel code block generated with Spoon | END
+}

--- a/src/com/google/android/gms/location/LastLocationRequest.aidl
+++ b/src/com/google/android/gms/location/LastLocationRequest.aidl
@@ -1,0 +1,3 @@
+package com.google.android.gms.location;
+
+parcelable LastLocationRequest;

--- a/src/com/google/android/gms/location/LastLocationRequest.java
+++ b/src/com/google/android/gms/location/LastLocationRequest.java
@@ -1,0 +1,45 @@
+package com.google.android.gms.location;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import app.grapheneos.gmscompat.safeparcel.Property;
+import app.grapheneos.gmscompat.safeparcel.SafeParcel;
+import app.grapheneos.gmscompat.safeparcel.SpReadOnly;
+
+// https://developers.google.com/android/reference/com/google/android/gms/location/LastLocationRequest
+public class LastLocationRequest extends SpReadOnly {
+    @Property(1) public long maxAge = Long.MAX_VALUE;
+    @Property(2) public int granularity = LocationRequest.GRANULARITY_PERMISSION_LEVEL;
+//    @Property(3) public boolean locationSettingsIgnored; // also named "bypass"
+//    @Property(4) public String moduleId;
+//    @Property(5) public ClientIdentity impersonation;
+
+// SafeParcel code block generated with Spoon | START
+    public static final Parcelable.Creator<LastLocationRequest> CREATOR = new Parcelable.Creator<LastLocationRequest>() {
+        public LastLocationRequest createFromParcel(Parcel p) {
+            LastLocationRequest o = new LastLocationRequest();
+            final int objectEnd = SafeParcel.getObjectEnd(p);
+            while (p.dataPosition() < objectEnd) {
+                int ph = SafeParcel.propHeader(p);
+                switch (SafeParcel.propId(ph)) {
+                    case 1 :
+                        o.maxAge = SafeParcel.readLong(ph, p);
+                        continue;
+                    case 2 :
+                        o.granularity = SafeParcel.readInt(ph, p);
+                        continue;
+                    default :
+                        SafeParcel.skipProp(ph, p);
+                }
+            }
+            SafeParcel.checkFullRead(objectEnd, p);
+            return o;
+        }
+
+        public LastLocationRequest[] newArray(int size) {
+            return new LastLocationRequest[size];
+        }
+    };
+// SafeParcel code block generated with Spoon | END
+}

--- a/src/com/google/android/gms/location/LocationAvailabilityRequest.aidl
+++ b/src/com/google/android/gms/location/LocationAvailabilityRequest.aidl
@@ -1,0 +1,3 @@
+package com.google.android.gms.location;
+
+parcelable LocationAvailabilityRequest;

--- a/src/com/google/android/gms/location/LocationAvailabilityRequest.java
+++ b/src/com/google/android/gms/location/LocationAvailabilityRequest.java
@@ -1,0 +1,29 @@
+package com.google.android.gms.location;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import app.grapheneos.gmscompat.safeparcel.SafeParcel;
+import app.grapheneos.gmscompat.safeparcel.SpReadOnly;
+
+public class LocationAvailabilityRequest extends SpReadOnly {
+//    @Property(1) public boolean bypass;
+//    @Property(2) public ClientIdentity impersonation;
+
+    public static final Parcelable.Creator<LocationAvailabilityRequest> CREATOR = new Creator<>() {
+        public LocationAvailabilityRequest createFromParcel(Parcel p) {
+            LocationAvailabilityRequest o = new LocationAvailabilityRequest();
+            final int objectEnd = SafeParcel.getObjectEnd(p);
+            while (p.dataPosition() < objectEnd) {
+                int ph = SafeParcel.propHeader(p);
+                SafeParcel.skipProp(ph, p);
+            }
+            SafeParcel.checkFullRead(objectEnd, p);
+            return o;
+        }
+
+        public LocationAvailabilityRequest[] newArray(int size) {
+            return new LocationAvailabilityRequest[size];
+        }
+    };
+}

--- a/src/com/google/android/gms/location/LocationRequest.java
+++ b/src/com/google/android/gms/location/LocationRequest.java
@@ -14,6 +14,15 @@ public class LocationRequest extends SpReadOnly {
     public static final int PRIORITY_LOW_POWER = 104;
     public static final int PRIORITY_NO_POWER = 105;
 
+    // https://developers.google.com/android/reference/com/google/android/gms/location/Granularity
+    public static final int GRANULARITY_PERMISSION_LEVEL = 0;
+    public static final int GRANULARITY_COARSE = 1;
+    public static final int GRANULARITY_FINE = 2;
+
+    public static final int THROTTLE_BACKGROUND = 0;
+    public static final int THROTTLE_ALWAYS = 1;
+    public static final int THROTTLE_NEVER = 2;
+
     @Property(1) public int priority;
     @Property(2) public long interval;
     @Property(3) public long minUpdateIntervalMillis;
@@ -23,12 +32,11 @@ public class LocationRequest extends SpReadOnly {
     @Property(8) public long maxUpdateDelayMillis;
     @Property(9) public boolean waitForAccurateLocation;
     @Property(10) public long durationMillis = Long.MAX_VALUE;
-
+    @Property(12) public int granularity = GRANULARITY_PERMISSION_LEVEL;
     /*
     unused for now:
 
     @Property(11) public long maxUpdateAgeMillis = -1L;
-    @Property(12) public int granularity;
     @Property(13) public int throttleBehavior;
     @Property(14) public String moduleId;
     @Property(15) public boolean bypass;
@@ -70,14 +78,17 @@ public class LocationRequest extends SpReadOnly {
                     case 10 :
                         o.durationMillis = SafeParcel.readLong(ph, p);
                         continue;
+                    case 12 :
+                        o.granularity = SafeParcel.readInt(ph, p);
+                        continue;
                     default :
                         SafeParcel.skipProp(ph, p);
                 }
-            } 
+            }
             SafeParcel.checkFullRead(objectEnd, p);
             return o;
         }
-    
+
         public LocationRequest[] newArray(int size) {
             return new LocationRequest[size];
         }

--- a/src/com/google/android/gms/location/LocationResult.java
+++ b/src/com/google/android/gms/location/LocationResult.java
@@ -21,7 +21,7 @@ public class LocationResult extends SpWriteOnly {
 // SafeParcel code block generated with Spoon | START
     public void writeToParcel(Parcel p, int wtpFlags) {
         final int headerEnd = SafeParcel.beginObjectHeader(p);
-        SafeParcel.writeParcelableInternal(1, this.locations, p, 0);
+        SafeParcel.writeParcelableList(1, this.locations, p, 0);
         SafeParcel.completeObjectHeader(headerEnd, p);
     }
 

--- a/src/com/google/android/gms/location/internal/IGoogleLocationManagerService.aidl
+++ b/src/com/google/android/gms/location/internal/IGoogleLocationManagerService.aidl
@@ -5,23 +5,40 @@ import android.location.Location;
 import android.os.Bundle;
 
 import com.google.android.gms.common.api.Status;
-import com.google.android.gms.location.internal.IFusedLocationProviderCallback;
-import com.google.android.gms.location.internal.ISettingsCallbacks;
-import com.google.android.gms.location.internal.LocationRequestInternal;
-import com.google.android.gms.location.internal.LocationRequestUpdateData;
+import com.google.android.gms.common.internal.ICancelToken;
+import com.google.android.gms.location.CurrentLocationRequest;
+import com.google.android.gms.location.LastLocationRequest;
 import com.google.android.gms.location.LocationAvailability;
+import com.google.android.gms.location.LocationAvailabilityRequest;
 import com.google.android.gms.location.LocationRequest;
 import com.google.android.gms.location.LocationSettingsRequest;
 import com.google.android.gms.location.ILocationListener;
+import com.google.android.gms.location.internal.IFusedLocationProviderCallback;
+import com.google.android.gms.location.internal.ILocationStatusCallback;
+import com.google.android.gms.location.internal.ISettingsCallbacks;
+import com.google.android.gms.location.internal.IStatusCallback;
+import com.google.android.gms.location.internal.LocationReceiver;
+import com.google.android.gms.location.internal.LocationRequestInternal;
+import com.google.android.gms.location.internal.LocationRequestUpdateData;
 
 interface IGoogleLocationManagerService {
     Location getLastLocation() = 6;
     Location getLastLocation2(String packageName) = 20;
     Location getLastLocation3(String contextAttributionTag) = 79;
+    void getLastLocation4(in LastLocationRequest request, ILocationStatusCallback callback) = 81;
+    void getLastLocation5(in LastLocationRequest request, in LocationReceiver receiver) = 89;
 
     LocationAvailability getLocationAvailability(String packageName) = 33;
+    void getLocationAvailability2(in LocationAvailabilityRequest request, in LocationReceiver receiver) = 90;
+
     void requestLocationSettingsDialog(in LocationSettingsRequest settingsRequest, ISettingsCallbacks callback, String packageName) = 62;
 
     void updateLocationRequest(in LocationRequestUpdateData locationRequestUpdateData) = 58;
     void flushLocations(IFusedLocationProviderCallback callback) = 66;
+
+    void registerLocationReceiver(in LocationReceiver receiver, in LocationRequest request, IStatusCallback callback) = 87;
+    void unregisterLocationReceiver(in LocationReceiver receiver, IStatusCallback callback) = 88;
+
+    ICancelToken getCurrentLocation(in CurrentLocationRequest request, ILocationStatusCallback callback) = 86;
+    ICancelToken getCurrentLocation2(in CurrentLocationRequest request, in LocationReceiver receiver) = 91;
 }

--- a/src/com/google/android/gms/location/internal/ILocationAvailabilityStatusCallback.aidl
+++ b/src/com/google/android/gms/location/internal/ILocationAvailabilityStatusCallback.aidl
@@ -1,0 +1,8 @@
+package com.google.android.gms.location.internal;
+
+import com.google.android.gms.common.api.Status;
+import com.google.android.gms.location.LocationAvailability;
+
+interface ILocationAvailabilityStatusCallback {
+    void onResult(in Status status, in LocationAvailability locationAvailability);
+}

--- a/src/com/google/android/gms/location/internal/ILocationStatusCallback.aidl
+++ b/src/com/google/android/gms/location/internal/ILocationStatusCallback.aidl
@@ -1,0 +1,9 @@
+package com.google.android.gms.location.internal;
+
+import android.location.Location;
+
+import com.google.android.gms.common.api.Status;
+
+interface ILocationStatusCallback {
+    void onResult(in Status status, in Location location);
+}

--- a/src/com/google/android/gms/location/internal/IStatusCallback.aidl
+++ b/src/com/google/android/gms/location/internal/IStatusCallback.aidl
@@ -1,0 +1,9 @@
+package com.google.android.gms.location.internal;
+
+import android.location.Location;
+
+import com.google.android.gms.common.api.Status;
+
+interface IStatusCallback {
+    void onCompletion(in Status status);
+}

--- a/src/com/google/android/gms/location/internal/LocationReceiver.aidl
+++ b/src/com/google/android/gms/location/internal/LocationReceiver.aidl
@@ -1,0 +1,3 @@
+package com.google.android.gms.location.internal;
+
+parcelable LocationReceiver;

--- a/src/com/google/android/gms/location/internal/LocationReceiver.java
+++ b/src/com/google/android/gms/location/internal/LocationReceiver.java
@@ -1,0 +1,56 @@
+package com.google.android.gms.location.internal;
+
+import android.app.PendingIntent;
+import android.os.IBinder;
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import app.grapheneos.gmscompat.safeparcel.Property;
+import app.grapheneos.gmscompat.safeparcel.SafeParcel;
+import app.grapheneos.gmscompat.safeparcel.SpReadOnly;
+
+public class LocationReceiver extends SpReadOnly {
+    public static final int TYPE_ILocationListener = 1;
+    public static final int TYPE_ILocationCallback = 2;
+    public static final int TYPE_PendingIntent = 3;
+    public static final int TYPE_ILocationStatusCallback = 4;
+    public static final int TYPE_ILocationAvailabilityStatusCallback = 5;
+
+    @Property(1) public int type;
+//    @Property(2) public IBinder;
+    @Property(3) public IBinder binder;
+    @Property(4) public PendingIntent pendingIntent;
+//    @Property(5) public String;
+//    @Property(6) public String;
+
+// SafeParcel code block generated with Spoon | START
+    public static final Parcelable.Creator<LocationReceiver> CREATOR = new Parcelable.Creator<LocationReceiver>() {
+        public LocationReceiver createFromParcel(Parcel p) {
+            LocationReceiver o = new LocationReceiver();
+            final int objectEnd = SafeParcel.getObjectEnd(p);
+            while (p.dataPosition() < objectEnd) {
+                int ph = SafeParcel.propHeader(p);
+                switch (SafeParcel.propId(ph)) {
+                    case 1 :
+                        o.type = SafeParcel.readInt(ph, p);
+                        continue;
+                    case 3 :
+                        o.binder = SafeParcel.readStrongBinder(ph, p);
+                        continue;
+                    case 4 :
+                        o.pendingIntent = SafeParcel.readParcelable(ph, p, PendingIntent.CREATOR);
+                        continue;
+                    default :
+                        SafeParcel.skipProp(ph, p);
+                }
+            }
+            SafeParcel.checkFullRead(objectEnd, p);
+            return o;
+        }
+
+        public LocationReceiver[] newArray(int size) {
+            return new LocationReceiver[size];
+        }
+    };
+// SafeParcel code block generated with Spoon | END
+}

